service/controller/resource/clusterreleaseversion/checkupgradeprogress.go

```diff
@@ -0,0 +1,65 @@
+package clusterreleaseversion
+
+import (
+	"context"
+	"time"
+
+	aeconditions "github.com/giantswarm/apiextensions/v3/pkg/conditions"
+	"github.com/giantswarm/microerror"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
+	capiconditions "sigs.k8s.io/cluster-api/util/conditions"
+
+	"github.com/giantswarm/azure-operator/v5/pkg/conditions"
+)
+
+func (r *Resource) isUpgradeCompleted(ctx context.Context, cluster *capi.Cluster) (bool, error) {
+	r.logger.LogCtx(ctx, "level", "debug", "message", "checking if cluster upgrade has been completed")
+	// Here we expect that Cluster Upgrading conditions is set to True.
+	upgradingCondition := capiconditions.Get(cluster, aeconditions.UpgradingCondition)
+	if !conditions.IsTrue(upgradingCondition) {
+		err := microerror.Maskf(
+			conditions.UnexpectedConditionStatusError,
+			conditions.ExpectedTrueErrorMessage(cluster, aeconditions.UpgradingCondition))
+		return false, err
+	}
+
+	// Upgrading is in progress, now let's check if it has been completed.
+
+	// But don't check if Upgrading has been completed for the first 5 minutes,
+	// give other controllers time to start reconciling their CRs.
+	if time.Now().Before(upgradingCondition.LastTransitionTime.Add(5 * time.Minute)) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "upgrade is in progress for less than 5 minutes, check back later")
+		return false, nil
+	}
+
+	// Cluster has been in Upgrading state for at least 5 minutes now, so
+	// let's check if it is Ready.
+	readyCondition := capiconditions.Get(cluster, capi.ReadyCondition)
+
+	if !conditions.IsTrue(readyCondition) {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "cluster not ready, upgrade is still in progress")
+		return false, nil
+	}
+
+	// (1) In addition to cluster being ready, here we check that it actually
+	// became ready during the upgrade, which would mean that the upgrade has
+	// been completed.
+	becameReadyWhileUpgrading := readyCondition.LastTransitionTime.After(upgradingCondition.LastTransitionTime.Time)
+
+	// (2) Or we declare Upgrading to be completed if nothing happened for 15
```

Comment:
it's a new comment and it got outdated already :joy: 
```suggestion
	// (2) Or we declare Upgrading to be completed if nothing happened for 20
```